src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,337 @@
+                //
+                // Note that we need to register separately for OperationKind.Invocation due to https://github.com/dotnet/roslyn/issues/26206
+
+                compilationStartContext.RegisterSymbolAction(compilationAnalyzer.AnalyzeSymbolDeclaration, SymbolKind.Method, SymbolKind.Field, SymbolKind.Property, SymbolKind.Event);
+
+                compilationStartContext.RegisterSymbolStartAction(symbolStartContext =>
+                {
+                    var hasInvalidOperation = false;
+                    symbolStartContext.RegisterOperationAction(compilationAnalyzer.AnalyzeMemberReferenceOperation, OperationKind.FieldReference, OperationKind.MethodReference, OperationKind.PropertyReference, OperationKind.EventReference);
+                    symbolStartContext.RegisterOperationAction(compilationAnalyzer.AnalyzeFieldInitializer, OperationKind.FieldInitializer);
+                    symbolStartContext.RegisterOperationAction(compilationAnalyzer.AnalyzeInvocationOperation, OperationKind.Invocation);
+                    symbolStartContext.RegisterOperationAction(_ => hasInvalidOperation = true, OperationKind.Invalid);
+                    symbolStartContext.RegisterSymbolEndAction(symbolEndContext => compilationAnalyzer.OnSymbolEnd(symbolEndContext, hasInvalidOperation));
+                }, SymbolKind.NamedType);
+            });
+        }
+
+        private sealed class CompilationAnalyzer
+        {
+            private readonly DiagnosticDescriptor _removeUnusedMembersRule, _removeUnreadMembersRule;
+            private readonly object _gate;
+            private readonly Dictionary<ISymbol, ValueUsageInfo> _symbolValueUsageStateMap;
+            private readonly Lazy<INamedTypeSymbol> _lazyTaskType, _lazyGenericTaskType;
+
+            public CompilationAnalyzer(Compilation compilation, DiagnosticDescriptor removeUnusedMembersRule, DiagnosticDescriptor removeUnreadMembersRule)
+            {
+                _removeUnusedMembersRule = removeUnusedMembersRule;
+                _removeUnreadMembersRule = removeUnreadMembersRule;
+                _gate = new object();
+
+                // State map for candidate member symbols, with the value indicating how each symbol is used in executable code.
+                _symbolValueUsageStateMap = new Dictionary<ISymbol, ValueUsageInfo>();
+
+                _lazyTaskType = new Lazy<INamedTypeSymbol>(() => compilation.TaskType());
+                _lazyGenericTaskType = new Lazy<INamedTypeSymbol>(() => compilation.TaskOfTType());
+            }
+
+            public void AnalyzeSymbolDeclaration(SymbolAnalysisContext symbolContext)
+            {
+                if (IsCandidateSymbol(symbolContext.Symbol))
+                {
+                    lock (_gate)
+                    {
+                        // Initialize unused state to 'true'.
```

Comment:
this doesn't match the code.  You intialize to 'None'.  Might be worth explaining what that means.